src/day.rs

```diff
@@ -0,0 +1,231 @@
+use core::cmp::Ordering;
+use core::convert::TryFrom;
+use core::fmt;
+use core::fmt::Debug;
+use core::fmt::Display;
+use core::fmt::Write;
+use core::ops::Add;
+use core::ops::Sub;
+
+use crate::oldtime;
+use crate::DateTime;
+use crate::Days;
+use crate::NaiveDate;
+use crate::NaiveTime;
+use crate::TimeZone;
+use oldtime::Duration;
+
+/// Represents the full range of local timestamps in the day
+///
+///
+#[derive(Clone, Copy)]
+pub struct Day<Tz>
+where
+    Tz: TimeZone + Copy + Display,
+{
+    date: NaiveDate,
+    tz: Tz,
+}
+
+impl<Tz> Debug for Day<Tz>
+where
+    Tz: TimeZone + Copy + Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(&self.date, f)?;
+        f.write_char(' ')?;
+        self.tz.fmt(f)
+    }
+}
+
+impl<Tz> Display for Day<Tz>
+where
+    Tz: TimeZone + Copy + Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+impl<Tz> PartialOrd for Day<Tz>
+where
+    Tz: TimeZone + Copy + Display,
+{
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.date.partial_cmp(&other.date)
+    }
+}
+
+impl<Tz> Ord for Day<Tz>
+where
+    Tz: TimeZone + Copy + Display,
+{
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.date.cmp(&other.date)
+    }
+}
+
+impl<Tz> PartialEq for Day<Tz>
+where
+    Tz: TimeZone + Copy + Display,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.date.eq(&other.date)
+    }
+}
+
+impl<Tz> Eq for Day<Tz> where Tz: TimeZone + Copy + Display {}
+
+impl<Tz> Day<Tz>
+where
+    Tz: TimeZone + Copy + Display,
+{
+    ///
+    pub fn date(&self) -> NaiveDate {
+        self.date
+    }
+
+    ///
+    pub fn zone(&self) -> Tz {
+        self.tz
+    }
+
+    ///
+    pub fn new(date: NaiveDate, tz: Tz) -> Day<Tz> {
+        Day { date, tz }
+    }
+
+    ///
+    pub fn succ(&self) -> Option<Day<Tz>> {
+        Some(Day { date: self.date.succ_opt()?, tz: self.tz })
+    }
+
+    ///
+    pub fn pred(&self) -> Option<Day<Tz>> {
+        Some(Day { date: self.date.pred_opt()?, tz: self.tz })
+    }
+
+    ///
+    pub fn start(&self) -> DateTime<Tz> {
```

Comment:
Could use some documentation of *why* we have to use a loop here. What problem is this trying to solve?

Reply:
I've added some more documentation here including around panics and incorrect results

Reply:
One way to have this not require heuristics would be to make the loop slightly differently:
- first, try with NaiveTime::MIN as a fast path
- then, for each 15min segment,
  - if MIN + segment_end is not ok, go to the next segment
  - else, if MIN + segment_end - 1ns is ok too, do a dichotomy between MIN + segment_begin and MIN + segment_end

That said, I personally don't think that doing this would be necessary. Summer time changes across midnight are unlikely enough already, having them not be a multiple-of-15-min offset would be actively harmful to basically any software.

Another idea would be, to delegate the implementation of this to a new trait method on TimeZone, that would have a default implementation of adding NaiveTime::MIN; and on the day chrono_tz gets a summer time transition across midnight the specialization can be added there.

I'm personally partial towards the second idea, as it gets us all of the simplest implementation right now, the most efficient one, and the most future-proof one. Does that make sense?